lib/format_erf.c

```diff
@@ -645,31 +645,157 @@ static int erf_read_packet(libtrace_t *libtrace, libtrace_packet_t *packet) {
 	return rlen;
 }
 
-static int erf_dump_packet(libtrace_out_t *libtrace,
-		dag_record_t *erfptr, int framinglen, void *buffer,
-                int caplen) {
-	int numbytes = 0;
+static bool find_compatible_linktype(libtrace_out_t *libtrace,
+                                libtrace_packet_t *packet)
+{
+        /* Keep trying to simplify the packet until we can find 
+         * something we can do with it */
+        do {
+                char type=libtrace_to_erf_type(trace_get_link_type(packet));
+
+                /* Success */
+                if (type != (char)-1)
+                        return true;
+
+                if (!demote_packet(packet)) {
+                        trace_set_err_out(libtrace,
+                                        TRACE_ERR_NO_CONVERSION,
+                                        "No erf type for packet (%i)",
+                                        trace_get_link_type(packet));
+                        return false;
+                }
+
+        } while(1);
+
+        return true;
+}
+
+int libtrace_to_erf_hdr(libtrace_out_t *libtrace, libtrace_packet_t *packet,
+    dag_record_t *erf, int *framinglen, int *caplen, int *padding) {
+
+    int wlen = trace_get_wire_length(packet);
+    *framinglen = dag_record_size + erf_get_padding(packet);
+    *caplen = trace_get_capture_length(packet);
+    *padding = 0;
+
+    if (*caplen <= 0 || *caplen > 65536) {
+        trace_set_err_out(libtrace, TRACE_ERR_BAD_PACKET,
+            "Capture length is out of range in libtrace_to_erf_hdr()");
+        return -1;
+    }
+
+    if (*framinglen > 65536) {
+        trace_set_err_out(libtrace, TRACE_ERR_BAD_PACKET,
+            "Framing length is to large in libtrace_to_erf_hdr()");
+        return -1;
+    }
+
+    if (*caplen + *framinglen <= 0 || *caplen + *framinglen > 65536) {
+        trace_set_err_out(libtrace, TRACE_ERR_BAD_PACKET,
+            "Capture + framing length is out of range in libtrace_to_erf_hdr()");
+        return -1;
+    }
+
+    erf->ts = bswap_host_to_le64(trace_get_erf_timestamp(packet));
+    memset(&erf->flags, 0, sizeof(erf->flags));
+    if (trace_get_direction(packet) != TRACE_DIR_UNKNOWN)
+        erf->flags.iface = trace_get_direction(packet);
+    if (!find_compatible_linktype(libtrace,packet))
+        return -1;
+    erf->type = libtrace_to_erf_type(trace_get_link_type(packet));
+
+    // calculate correct padding for the ERF record
+    erf_calculate_padding(packet, framinglen, caplen, padding);
+
+    erf->rlen = htons(*framinglen + *caplen + *padding);
+    // loss counter. Cant't do this
+    erf->lctr = 0;
+    erf->wlen = htons(wlen);
+
+    return 0;
+}
+
+static inline int erf_increased_padding(int framinglen, int caplen) {
+    int padding;
+    padding = sizeof(uint64_t) - ((framinglen + caplen) % sizeof(uint64_t));
+    if (padding == sizeof(uint64_t))
```

Comment:
The compiler will be able to optimise
padding = (alignment - (value % alignment)) % alignment
better as alignment is almost always a multiple of base two and the modulus operations get replaced with bit-masking.